docs: User documentation on plugin manager (#69)
Bit-Quill/aws-mysql-jdbc: README.md

  <dependency>
    <groupId>software.aws.rds</groupId>
    <artifactId>aws-mysql-jdbc</artifactId>

>     <version>0.4.0</version>

Reviewer: Need to update the Gradle dependency as well

> 

Reviewer: Can you please fix a typo in the README
```
| RDS Custom Cluster | `jdbc:mysql:aws://db-identifier.cluster-custom-XYZ.us-east-2l.rds.amazonaws.com:3306`      |    None | *Initial connection:* any DB instance in the custom DB cluster<br/>*Failover 
```
Should be `us-east-2` instead of `us-east-2l`

In addition to [the parameters that can be configured for the MySQL Connector/J driver](https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-configuration-properties.html), the following parameters can also be passed to the AWS JDBC Driver through the connection URL to configure additional driver behavior.

| Parameter       | Value           | Required      | Description  | Default Value |
| ------------- |:-------------:|:-------------:|:-------------:| --------- |
|`enableClusterAwareFailover` | Boolean | No | Set to true to enable the fast failover behavior offered by the AWS JDBC Driver. Set to false for simple JDBC connections that do not require fast failover functionality. | `true` |

> |`clusterInstanceHostPattern` | String | If connecting using an IP address or custom domain URL: Yes<br/>Otherwise: No | This parameter is not required unless connecting to an AWS RDS cluster via an IP address or custom domain URL. In those cases, this parameter specifies the cluster instance DNS pattern that will be used to build a complete instance endpoint. A "?" character in this pattern should be used as a placeholder for the DB instance identifiers of the instances in the cluster. <br/><br/>Example: `?.my-domain.com`, `any-subdomain.?.my-domain.com:9999`<br/><br/>Usecase Example: If your cluster instance endpoints followed this pattern:`instanceIdentifier1.customHost`, `instanceIdentifier2.customHost`, etc. and you wanted your initial connection to be to `customHost:1234`, then your connection string should look something like this: `jdbc:mysql:aws://customHost:1234/test?clusterInstanceHostPattern=?.customHost` | If unspecified, and the provided connection string is not an IP address or custom domain, the driver will automatically acquire the cluster instance host pattern from the customer-provided connection string. |

Reviewer 2: I feel like default value already implies the absence of the property or value. Do we need to say "If unspecified...". `enableClusterAwareFailover`, for example ,if unspecified takes on the value of true, but we don't need to preface that with "If unspecified".

I think we can chop off some wording here and just go straight to the default value. e.g., "The cluster instance host pattern."

Similar comment for `clusterId` as well.

> |`clusterId` | String | No | A unique identifier for the cluster. Connections with the same cluster id share a cluster topology cache. This connection parameter is not required and thus should only be set if desired. | If unspecified, the driver will automatically acquire a cluster id for AWS RDS clusters. |

Reviewer 2: ID should be in caps in the `required` section.

|`clusterTopologyRefreshRateMs` | Integer | No | Cluster topology refresh rate in milliseconds. The cached topology for the cluster will be invalidated after the specified time, after which it will be updated during the next interaction with the connection. | `30000` |
|`failoverTimeoutMs` | Integer | No | Maximum allowed time in milliseconds to attempt reconnecting to a new writer or reader instance after a cluster failover is initiated. | `60000` |
|`failoverClusterTopologyRefreshRateMs` | Integer | No | Cluster topology refresh rate in milliseconds during a writer failover process. During the writer failover process, cluster topology may be refreshed at a faster pace than normal to speed up discovery of the newly promoted writer. | `5000` |
|`failoverWriterReconnectIntervalMs` | Integer | No | Interval of time in milliseconds to wait between attempts to reconnect to a failed writer during a writer failover process. | `5000` |
|`failoverReaderConnectTimeoutMs` | Integer | No | Maximum allowed time in milliseconds to attempt to connect to a reader instance during a reader failover process. | `5000` |

> |`acceptAwsProtocolOnly` | Boolean | If using simultaneously with another MySQL driver that supports the same protocols: Yes<br/>Otherwise: No | Set to true to only accept connections for URLs with the jdbc:mysql:aws:// protocol. This setting should be set to true when running an application that uses this driver simultaneously with another MySQL driver that supports the same protocols (eg the MySQL JDBC Driver), to ensure the driver protocols do not clash. This behavior can also be set at the driver level for every connection via the Driver.setAcceptAwsProtocolOnly method; however, this connection parameter will take priority when present. | `false` |

Reviewer: ```suggestion
|`acceptAwsProtocolOnly` | Boolean | If using simultaneously with another MySQL driver that supports the same protocols: Yes<br/>Otherwise: No | Set to true to only accept connections for URLs with the jdbc:mysql:aws:// protocol. This setting should be set to true when running an application that uses this driver simultaneously with another MySQL driver that supports the same protocols (e.g. the MySQL JDBC Driver), to ensure the driver protocols do not clash. This behavior can also be set at the driver level for every connection via the Driver.setAcceptAwsProtocolOnly method; however, this connection parameter will take priority when present. | `false` |
```

Reviewer: I think we should left-align the `Description` column as well.

 * Scenario 1: Failover happens when autocommit is set to true - Catch SQLException with code 08S02.
 */
public class FailoverSampleApp1 {

>    private static final String CONNECTION_STRING = "jdbc:mysql:aws://database-mysql.cluster-XYZ.us-east-2.rds.amazonaws.com:3306/employees";

Reviewer: Oh no 😱 , I might have copy pasted the code with wrong spacing. Can you please double check the spacing?

|`failureDetectionCount` | Integer | No | Number of failed connection checks before considering database node as unhealthy. | `3` |
|`monitorDisposalTime` | Integer | No | Interval in milliseconds for a monitor to be considered inactive and to be disposed. | `60000` |

## Extra Additions

### XML Entity Injection Fix

The default XML parser contained a security risk which made the driver prone to XXE (XML Entity Injection) attacks. To solve this issue, we disabled DTDs (Document type definition) in the XML parser. If you require this restriction to be lifted in your application, you can set the `allowXmlUnsafeExternalEntity` parameter in the connection string to true. Please see table below for a definition of this parameter. 

| Parameter       | Value           | Required      | Description  | Default Value |
| ------------- |:-------------:|:-------------:|:-------------:| --------- |
|`allowXmlUnsafeExternalEntity` | Boolean | No | Set to true if you would like to use XML inputs that refer to external entities. WARNING: Setting this to true is unsafe since your system to be prone to XXE attacks. | `false` |

### AWS IAM Database Authentication

> An optional authentication method is to use Amazon AWS Identity and Access Management (IAM). Use of IAM requires RDS SDK to be added to the classpath.

Reviewer: https://github.com/awslabs/aws-mysql-jdbc#aws-iam-database-authentication
I like this version more. Maybe we should leave it untouched or update main and dev with the latest changes from upstream.

>2. It is highly recommended that you use the cluster and read-only cluster endpoints instead of the direct instance endpoints of your Aurora cluster, unless you are confident about your application's usage of instance endpoints. Although the Driver will correctly failover to the new writer instance when using instance endpoints, usage of these endpoints are discouraged because individual instances can spontaneously change reader/writer status when failover occurs. The driver will always connect directly to the instance specified if an instance endpoint is provided, so a write-safe connection cannot be assumed if the application uses instance endpoints.

### Connection Plugin Manager

> Connection plugin manager initializes, triggers, and cleans up connection plugins. Connection plugins are widgets attached to each `Connection` objects to help execute additional or supplementary logic related to that `Connection`. [Enhanced Failure Monitoring](https://github.com/awslabs/aws-mysql-jdbc#enhanced-failure-monitoring) is one example of a connection plugin. 

Reviewer: ```suggestion
Connection plugin manager initializes, triggers, and cleans up a chain of connection plugins. Connection plugins are widgets attached to each `Connection` objects to help execute additional or supplementary logic related to that `Connection`. [Enhanced Failure Monitoring](https://github.com/awslabs/aws-mysql-jdbc#enhanced-failure-monitoring) is one example of a connection plugin. 
```

latter plugins. If the connection plugin at the head of the connection plugin chain measures the
execution time, this measurement would encompass the time spent in all the connection plugins down
the chain.

To learn how to write custom plugins, refer to examples located inside [Custom Plugins Demo](https://github.com/awslabs/aws-mysql-jdbc/tree/main/src/demo/java/customplugins).

#### Connection Plugin Manager Parameters
| Parameter       | Value           | Required      | Description  | Default Value |
| ------------- |:-------------:|:-------------:|:-------------:| --------- |
|`connectionPluginFactories` | String | No | String of fully-qualified class name of plugin factories. <br/><br/>Each factory in the string should be comma-separated `,`<br/><br/>**NOTE: The order of factories declared matters.**  <br/><br/>Example: `customplugins.MethodCountConnectionPluginFactory`, `customplugins.MethodCountConnectionPluginFactory,customplugins.ExecutionTimeConnectionPluginFactory,com.mysql.cj.jdbc.ha.ca.plugins.NodeMonitoringConnectionPluginFactory` | `com.mysql.cj.jdbc.ha.ca.plugins.NodeMonitoringConnectionPluginFactory` |

> >### :warning: Warnings About Connection Plugin Manager

Reviewer: I don't think we need `Warnings About...` after `:warning:`

Reviewer: Can we use a more descriptive title? Similarly for the other warning

Author: `:warning:` turns into ⚠️

> >If application is unable to close, it may be due to connection plugin manager not being able to release resources properly. User can manually release resources by calling `software.aws.rds.jdbc.mysql.Driver.releasePluginManagers()`.

Reviewer: ```suggestion
>If the application is unable to close, it may be due to connection plugin manager not being able to release resources properly. User can manually release resources by calling `software.aws.rds.jdbc.mysql.Driver.releasePluginManagers()`.
```

### Enhanced Failure Monitoring
<div style="text-align:center"><img src="./docs/files/images/enhanced_failure_monitoring_diagram.png" /></div>
The figure above shows a simplified workflow of enhanced failure monitoring. Enhanced failure monitoring, is a connection plugin implemented by using a monitor thread. The monitor will periodically check the connected database node's health. In the case of the database node showing up as unhealthy, the query will be retried with a new database node and the monitor is restarted. 

> Enhanced Failure Monitoring is loaded in by default and can be disabled by setting parameter `failureDetectionEnabled` to `false`. 

Reviewer: We sometimes use `Enhanced Failure Monitoring` and sometimes use `enhanced failure monitoring`. We should be more consistent, do we want to consider this a proper noun or not?

  }
}
```

> ### Plugin Manager

Reviewer 3: Connection Plugins?

<div style="text-align:center"><img src="./docs/files/images/plugin_manager_diagram.png" /></div>

> Figure shows a simplified workflow of the plugin manager. Starting at the top, when a new query is executed by the application, it is passed into the plugin manager. From the  plugin manager, the query is passed in order of plugins loaded, in this figure, `Custom Plugin A` to `Custom Plugin B` and finally to `Default Plugin` which executes the query and returns the result sets back up the chain.

Reviewer: Not only a query goes through the plugin manager.
I think this section should be called Connection Plugin Manager

Reviewer 3: Plugins execute Jdbc methods.

Custom plugins can be used to interact with queries, such as adding a feature to monitor a connection's status in the case of `NodeMonitoringConnectionPlugin`. 

> By default, `NodeMonitoringConnectionPlugin` is loaded for [Enhanced Failure Monitoring](https://github.com/awslabs/aws-mysql-jdbc#aws-iam-database-authentication). Additional custom plugins can be included and chained together.

Reviewer: Is this the correct hyperlink?

Reviewer 3: Additional custom plugins can be included and chained together. -> Additional custom plugins can be implemented  and used alongside to existing ones. Plugins can be chained together in a desired order.

Author: They're not supposed to work until it has been merged into AWS main

Reviewer: But the link is pointing to `aws-iam-database-authentication`

Author: Ah, good catch.

> To learn how to write custom plugins, refer to examples located inside [Custom Plugins Demo](https://github.com/awslabs/aws-mysql-jdbc/tree/main/src/demo/java/demo/customplugins).

Reviewer: ```suggestion
To learn how to write custom plugins, refer to examples located inside [Custom Plugins Demo](https://github.com/awslabs/aws-mysql-jdbc/tree/main/src/demo/java/customplugins).
```

**Note**: The order of which plugins load in matters.

#### Plugin Manager Parameters
| Parameter       | Value           | Required      | Description  |
| ------------- |:-------------:|:-------------:| ----- |

> |`connectionPluginFactories` | String | No | String of custom plugins to use monitoring. If left unset, `NodeMonitoringConnectionPluginFactory` will be loaded. <br/><br/>Each factory in the string should be separated by `,`<br/><br/>NOTE: The order of factories declared matters.  <br/><br/>Example `MyPluginA`, `MyPluginA,MyPluginB,NodeMonitoringConnectionPluginFactory` <br/><br/>**Default value:** `NodeMonitoringConnectionPluginFactory` |

Reviewer: You need to pass in the fully-qualified class name of the factory classes, not the plugins.

Reviewer 3: Need to explain what is default factory list in case user omits this parameter. And need to explain behavior for parameter with blank value.

> ### Enhanced Failure Monitoring

Reviewer: I thought the feature is called `Enhanced Instance Monitoring`. Can we confirm now so we stay consistent? @sergiyv-bitquill

Reviewer 3: Agree, we need a consistent name

Reviewer: I think Enhanced Failure Monitoring is more obvious, `instance` might not be immediately clear. Should we stick to `Enhanced Failure Monitoring`?

Reviewer: Can we add an additional column for default values instead of putting them in description?

Reviewer: While I understand that we want to keep our examples short and concise, we should aim to provide working examples as much as we can because it would be a better user experience. As in if we tell users they need to provide the `String of fully-qualified class name of plugin factories. `, our example should use fully-qualified class names.

`customplugins.MethodCountConnectionPluginFactory,customplugins.ExecutionTimeConnectionPluginFactory,com.mysql.cj.jdbc.ha.ca.plugins.NodeMonitoringConnectionPluginFactory`

Reviewer: I would just put `com.mysql.cj.jdbc.ha.ca.plugins.NodeMonitoringConnectionPluginFactory` for the default value.

> Enhanced failure monitoring, part of `NodeMonitoringConnectionPlugin`, periodically pings the database node to check if the database node is still healthy/available to be used. In the case of the database node showing up as downed/unhealthy, `NodeMonitoringConnectionPlugin` will automatically failover to a different database node.   

Reviewer 3: Enhanced failure monitoring, part of -> Enhanced failure monitoring, implemented by ...
pings -> checks health status

Reviewer: we should pick one between "healthy/available" and between "downed/unhealthy"

#### Enhanced Failure Monitoring Parameters
| Parameter       | Value           | Required      | Description  |
| ------------- |:-------------:|:-------------:| ----- |

> |`nativeFailureDetectionEnabled` | Boolean | No | Set to false if you would like to disable enhanced failure monitoring feature. <br/><br/>**Default value:** `true` |

Reviewer 3: Need to rename "nativeFailureDetectionEnabled" to be consistent with other parameters. Let's say "failureDetectionEnabled"

> |`failureDetectionTime` | Integer | No | Interval in milliseconds between sending a SQL query to the server and the first probe to the database node. <br/><br/>**Default value:** `30000` |

Reviewer 3: These 3 parameters are similar to TCP Keep Alive parameters and I think it'd be nice to mention that.

|`failureDetectionInterval` | Integer | No | Interval in milliseconds between probes to database node. <br/><br/>**Default value:** `5000` |
|`failureDetectionCount` | Integer | No | Number of failed connection checks before considering database node as unhealthy. <br/><br/>**Default value:** `3` |

> |`monitorDisposalTime` | Integer | No | Interval in milliseconds for a monitor to be considered inactive and to be disposed. <br/><br/>**Default value:** `60000` |

Reviewer: You can pass additional monitoring configurations by attaching `monitoring-` to the configuration key.

Reviewer 3: Need to explain what is monitor. Or to to avoid this word.

> >### :warning: Warnings About Proper Usage of the AWS JDBC Driver for MySQL

Reviewer 3: This section shouldn't be separated from failover section.

>2. It is highly recommended that you use the cluster and read-only cluster endpoints instead of the direct instance endpoints of your Aurora cluster, unless you are confident about your application's usage of instance endpoints. Although the Driver will correctly failover to the new writer instance when using instance endpoints, usage of these endpoints are discouraged because individual instances can spontaneously change reader/writer status when failover occurs. The driver will always connect directly to the instance specified if an instance endpoint is provided, so a write-safe connection cannot be assumed if the application uses instance endpoints.

### Connection Plugin Manager
<div style="text-align:center"><img src="./docs/files/images/connection_plugin_manager_diagram.png" /></div>

The figure above shows a simplified workflow of the connection plugin manager. 

> Starting at the top, when a JDBC method is executed by the driver, it is passed to the connection plugin manager. From the connection plugin manager, the JDBC method is passed in order of plugins loaded, in this figure, `Custom Plugin A` to `Custom Plugin B` and finally to `Default Plugin` which executes the JDBC method and returns the result sets back up the chain. Custom plugins can be used to interact with JDBC methods, such as adding a feature to monitor a connection's status in the case of `NodeMonitoringConnectionPlugin`.

Reviewer 3: We need general sentence what is plugin manager and what it is for. You can steal it from plugin sample page.

> By default, `NodeMonitoringConnectionPlugin` is loaded. Additional custom plugins can be implemented and used alongside existing ones. Plugins can be chained together in a desired order. Loading custom plugins will not include `NodeMonitoringConnectionPlugin` unless explicitly stated through the `connectionPluginFactories` parameter.

Reviewer 3: As discussed before, let's remove any references to NodeMonitoringConnectionPlugin. There's another chapter for it.

To learn how to write custom plugins, refer to examples located inside [Custom Plugins Demo](https://github.com/awslabs/aws-mysql-jdbc/tree/main/src/demo/java/customplugins).

#### Connection Plugin Manager Parameters
| Parameter       | Value           | Required      | Description  | Default Value |
| ------------- |:-------------:|:-------------:|:-------------:| --------- |
|`connectionPluginFactories` | String | No | String of fully-qualified class name of plugin factories. <br/><br/>Each factory in the string should be comma-separated `,`<br/><br/>**NOTE: The order of factories declared matters.**  <br/><br/>Example: `MyPluginAFactory`, `MyPluginAFactory,NodeMonitoringConnectionPluginFactory,MyPluginBFactory` | If unspecified or blank, `NodeMonitoringConnectionPluginFactory` will be loaded.|

### Enhanced Failure Monitoring

> <div style="text-align:center"><img src="./docs/files/images/node_monitoring_plugin_diagram.png" /></div>

Reviewer 3: Need to remove references to failover from the diagram. Also remove "ping" from it.

> The figure above shows a simplified workflow of enhanced failure monitoring. Enhanced failure monitoring, implemented by using a monitor thread, periodically pings the database node to check the node's health. In the case of the database node showing up as unhealthy, failover to a different database node occurs, restarts the monitor and retries the query.

Reviewer 3: Need an opening sentence that this's actually a plugin and it's goal is ...

Reviewer 3: I'd avoid word "ping". A better one is "check connection health status".

>   }

Reviewer: FailoverSampleApp1

```java
import java.sql.*;

/**
 * Scenario 1: Failover happens when autocommit is set to true - Catch SQLException with code 08S02.
 */
public class Main {
  private static final String CONNECTION_STRING = "jdbc:mysql:aws://database-mysql.cluster-XYZ.us-east-2.rds.amazonaws.com:3306/myDb";
  private static final String USERNAME = "username";
  private static final String PASSWORD = "password";
  private static final int MAX_RETRIES = 5;

  public static void main(String[] args) throws SQLException {
    // Create a connection.
    try (Connection conn = DriverManager.getConnection(CONNECTION_STRING, USERNAME, PASSWORD)) {
      // Configure the connection.
      setInitialSessionState(conn);

      // Do something with method "betterExecuteQuery" using the Cluster-Aware Driver.
      String select_sql = "SELECT * FROM employees";
      try (ResultSet rs = betterExecuteQuery(conn, select_sql)) {
        while (rs.next()) {
          System.out.println(rs.getString("first_name"));
        }
      }
    }
  }

  private static void setInitialSessionState(Connection conn) throws SQLException {
    // Your code here for the initial connection setup.
    try (Statement stmt1 = conn.createStatement()) {
      stmt1.executeUpdate("SET time_zone = \"+00:00\"");
    }
  }

  // A better executing query method when autocommit is set as the default value - True.
  private static ResultSet betterExecuteQuery(Connection conn, String query) throws SQLException {
    // Record the times of re-try.
    int retries = 0;

    while (true) {
      try {
        Statement stmt = conn.createStatement();
        return stmt.executeQuery(query);
      } catch (SQLException e) {
        // If the attempt to connect has failed MAX_RETRIES times,
        // throw the exception to inform users of the failed connection.
        if (retries > MAX_RETRIES) {
          throw e;
        }

        // Failover has occurred and the driver has failed over to another instance successfully.
        if (e.getSQLState().equalsIgnoreCase("08S02")) {
          // Re-config the connection.
          setInitialSessionState(conn);
          // Re-execute that query again.
          retries++;

        } else {
          // If some other exception occurs, throw the exception.
          throw e;
        }
      }
    }
  }
}
```

>2. It is highly recommended that you use the cluster and read-only cluster endpoints instead of the direct instance endpoints of your Aurora cluster, unless you are confident about your application's usage of instance endpoints. Although the Driver will correctly failover to the new writer instance when using instance endpoints, usage of these endpoints are discouraged because individual instances can spontaneously change reader/writer status when failover occurs. The driver will always connect directly to the instance specified if an instance endpoint is provided, so a write-safe connection cannot be assumed if the application uses instance endpoints.

> ### Connection Plugin Manager

Reviewer 2: I think we may need to rework this section a little bit. It needs a more structured introduction. Most customers looking into this probably don't know what a Connection Plugin Manager is. Consider answering a few of these questions as an outline:

- What's a connection plugin manager? Try to say more than just saying it's a manager for connection plugins. Without context, a question that may come to mind is how is failover associated with a connection plugin?
- What can I do with a connection plugin?
- Why would I need a connection plugin?

If you can give an overview of what a connection plugin manager is and tell me the value it brings, I would be more inclined to learn more about it.

Less is more and I would lead with the diagram and let that do the heavy lifting after the introduction. You can then walk through the diagram in the next paragraph for clarity. Going straight to connections are widgets and how they are chained might be too much, too soon. Picture first and then the explanation. We can probably combine some of this explanations with your explanation below the picture.

Connection Plugin Manager allows users to create and use connection plugins.

Connection plugins are widgets attached to each `Connection` objects to help execute additional or
supplementary logic related to that `Connection`. All the connection plugins are chained together,
where the prior connection plugin calls the next plugin. The AWS JDBC Driver for MySQL attaches the
`DefaultConnectionPlugin` to the tail of the connection plugin chain and actually executes the given
JDBC method.

Since all the connection plugins are chained together, the prior connection plugin affects the
latter plugins. If the connection plugin at the head of the connection plugin chain measures the
execution time, this measurement would encompass the time spent in all the connection plugins down
the chain.

<div style="text-align:center"><img src="./docs/files/images/connection_plugin_manager_diagram.png" /></div>

The figure above shows a simplified workflow of the connection plugin manager. 

> Starting at the top, when a JDBC method is executed by the driver, it is passed to the connection plugin manager. From the connection plugin manager, the JDBC method is passed in order of plugins loaded, in this figure, `Custom Plugin A` to `Custom Plugin B` and finally to `Default Plugin` which executes the JDBC method and returns the result sets back up the chain.

Reviewer: ```suggestion
Starting at the top, when a JDBC method is executed by the driver, it is passed to the connection plugin manager. From the connection plugin manager, the JDBC method is passed in order of plugins loaded, in this figure, `Custom Plugin A` to `Custom Plugin B` and finally to `Default Plugin` which executes the JDBC method and returns the result back up the chain.
```

#### Connection Plugin Manager Parameters
| Parameter       | Value           | Required      | Description  | Default Value |
| ------------- |:-------------:|:-------------:|:-------------:| --------- |
|`connectionPluginFactories` | String | No | String of fully-qualified class name of plugin factories. <br/><br/>Each factory in the string should be comma-separated `,`<br/><br/>**NOTE: The order of factories declared matters.**  <br/><br/>Example: `MyPluginAFactory`, `MyPluginAFactory,NodeMonitoringConnectionPluginFactory,MyPluginBFactory` | If unspecified or blank, Enhanced Failure Monitoring will be loaded.|

### Enhanced Failure Monitoring
Enhanced failure monitoring, is a connection plugin implemented by using a monitor thread. The monitor will periodically check the connected database node's health. In the case of the database node showing up as unhealthy, the query will be retried with a new database node and the monitor is restarted. 

Enhanced Failure Monitoring is loaded in by default and can be disabled by setting parameter `failureDetectionEnabled` to `false`. 

> If custom connection plugins are loaded, Enhanced Failure Monitoring will NOT be loaded unless explicitly included by adding `NodeMonitoringConnectionPluginFactory` to `connectionPluginFactories`.

Reviewer: ```suggestion
If custom connection plugins are loaded, Enhanced Failure Monitoring will NOT be loaded unless explicitly included by adding `com.mysql.cj.jdbc.ha.ca.plugins.NodeMonitoringConnectionPluginFactory` to `connectionPluginFactories`.
```

Reviewer: If I were the user I would be very confused if I copy what the README told me to do and my program throws an exception.

<div style="text-align:center"><img src="./docs/files/images/enhanced_failure_monitoring_diagram.png" /></div>
The figure above shows a simplified workflow of enhanced failure monitoring. 

#### Enhanced Failure Monitoring Parameters
`failureDetectionTime`, `failureDetectionInterval`, and `failureDetectionCount` are similar to TCP Keep Alive parameters.

Additional monitoring configurations can be included by adding the prefix `monitoring-` to the configuration key.

| Parameter       | Value           | Required      | Description  | Default Value |
| ------------- |:-------------:|:-------------:|:-------------:| --------- |

> |`failureDetectionEnabled` | Boolean | No | Set to false if you would like to disable enhanced failure monitoring feature. | `true` |

Reviewer: ```suggestion
|`failureDetectionEnabled` | Boolean | No | Set to false to disable enhanced failure monitoring feature. | `true` |
```

Reviewer: Normally we should avoid using second person in README

   1. Connect to MySQL database using master logins and create a new user with the following<br>
   `CREATE USER example_user_name IDENTIFIED WITH AWSAuthenticationPlugin AS 'RDS';`

| Parameter       | Value           | Required      | Description  | Default Value |
| ------------- |:-------------:|:-------------:|:-------------:| --------- |
|`useAwsIam` | Boolean | No | Set to true to enable AWS IAM database authentication | `false` |

###### Sample Code
```java
import java.sql.*;
import java.util.Properties;

public class AwsIamAuthenticationSample {

>    private static final String CONNECTION_STRING = "jdbc:mysql:aws://database-mysql-name.cluster-XYZ.us-east-2.rds.amazonaws.com:3306/employees";

Reviewer: Do the databases have to be in us-east-2? @matthewh-BQ I think we should be more generic.

   1. Connect to MySQL database using master logins and create a new user with the following<br>
   `CREATE USER example_user_name IDENTIFIED WITH AWSAuthenticationPlugin AS 'RDS';`

| Parameter       | Value           | Required      | Description  | Default Value |
| ------------- |:-------------:|:-------------:|:-------------:| --------- |

> |`useAwsIam` | Boolean | No | Set to true if you would like to use AWS IAM database authentication | `false` |

Reviewer: ```suggestion
|`useAwsIam` | Boolean | No | Set to true to enable AWS IAM database authentication | `false` |
```

> ###### Sample Code

Reviewer: ```java
import java.sql.*;
import java.util.Properties;

public class AwsIamAuthenticationSample {
  private static final String CONNECTION_STRING = "jdbc:mysql:aws://database-mysql-name.cluster-XYZ.us-east-2.rds.amazonaws.com:3306/employees";
  private static final String USER = "username";

  public static void main(String[] args) throws SQLException {
    final Properties properties = new Properties();
    // Enable AWS IAM database authentication
    properties.setProperty("useAwsIam", "true");
    properties.setProperty("user", USER);

    // Try and make a connection
    try (final Connection conn = DriverManager.getConnection(CONNECTION_STRING, properties);
        final Statement myQuery = conn.createStatement()) {
        try (final ResultSet rs = myQuery.executeQuery("SELECT * FROM employees")) {
          while (rs.next()) {
            System.out.println(rs.getString(1));
          }
        
      }
    }
  }
}
```

Reviewer: Also need to mention adding RDS SDK to the classpath
